vocode/streaming/audio/base_audio_service.py

```diff
@@ -0,0 +1,112 @@
+from vocode.streaming.models.audio_encoding import AudioEncoding
+from vocode.streaming.models.audio import AudioServiceConfig
+
+
+AudioServiceConfigType = TypeVar("AudioServiceConfigType", bound=AudioServiceConfig)
+
+
+class AbstractAudioService(Generic[AudioServiceConfigType]):
+    """Audio service abstract"""
+
+    def __init__(self, audio_service_config: AudioServiceConfigType):
+        self.audio_service_config = audio_service_config
+        self.is_muted = False
+
+    def mute(self):
+        """Mute"""
+        self.is_muted = True
+
+    def unmute(self):
+        """Unmute"""
+        self.is_muted = False
+
+    def get_audio_service_config(self) -> AudioServiceConfigType:
+        """get audio service config"""
+        return self.audio_service_config
+
+    async def ready(self):
+        """If audio service is ready"""
+        return True
+
+    def create_silent_chunk(self, chunk_size, sample_width=2):
+        """creating silence"""
+        linear_audio = b"\0" * chunk_size
+        if self.get_audio_service_config().audio_encoding == AudioEncoding.LINEAR16:
+            return linear_audio
+
+        if self.get_audio_service_config().audio_encoding == AudioEncoding.MULAW:
+            return audioop.lin2ulaw(linear_audio, sample_width)
+
+        raise ValueError(
+            f'"{self.get_audio_service_config().audio_encoding}" is not supported!'
+        )
+
+
+class BaseThreadAsyncAudioService(
+    AbstractAudioService[AudioServiceConfigType], ThreadAsyncWorker
+):
+    """Audio service"""
+
+    def __init__(
+        self,
+        conversation_id: str,
+        audio_service_config: AudioServiceConfigType,
+        logger: Optional[logging.Logger] = None,
+    ):
+        self.conversation_id = conversation_id
+        self.is_muted = False
+        self._ended = False
+        self.input_queue: asyncio.Queue[bytes] = asyncio.Queue()
+        self.output_queue: asyncio.Queue[bytes] = asyncio.Queue()
+        ThreadAsyncWorker.__init__(self, self.input_queue, self.output_queue)
+        AbstractAudioService.__init__(self, audio_service_config)
+        self.logger = logger
+
+    def process(self, chunk: bytes) -> bytes:
+        raise NotImplementedError
+
+    def _run_loop(self):
+        raise NotImplementedError
+
+    def send_audio(self, chunk):
+        """sending audio"""
+        if not self.is_muted:
+            self.consume_nonblocking(chunk)
+        else:
+            self.consume_nonblocking(self.create_silent_chunk(len(chunk)))
+
+    def generator(self):
+        """audio frame generator"""
+        while not self._ended:
+            try:
+                chunk = self.input_janus_queue.sync_q.get(timeout=5)
+            except queue.Empty:
+                return
+
+            if chunk is None:
+                return
+
+            data = [chunk]
+            # Now consume whatever other data's still buffered.
+            while True:
```

Comment:
what's the call pattern for `terminate()`? if `terminate()` is called in another thread while we're in the `while True` should we exit without continuing to read from the buffer? if so, we might want this to be also be `while not self._ended`

Reply:
`terminate()` is called via `streaming_conversation.py -> terminate()`. 

`input_janus_queue` is internal buffer which grab audio from `input_queue` which is running on main thread. As we terminate(), we stop sending audio to `input_queue` but we need get all audio already in `input_janus_queue` and process (these are frames already queued for being processed.